File: 卡码/代码随想录/1/06-二叉树/Python/06-二叉搜索树修改与构造/LeetCode_0450.py
#450. 删除二叉搜索树中的节点
from typing import *
from ..TypeNode import *

class Solution:
  def deleteNode(self, root: Optional[TreeNode], key: int) -> Optional[TreeNode]:
    if not root:
      return None
    if key<root.val:
      root.left=self.deleteNode(root.left,key)
    elif key>root.val:
      root.right=self.deleteNode(root.right,key)
    else:
      if not root.left and not root.right:
        return None
      elif root.left and not root.right:
        return root.left
      elif not root.left and root.right:
        return root.right
      else:
        parent:TreeNode=root
        tmp:TreeNode=root.right
        while tmp.left:
          parent=tmp
          tmp=tmp.left
        root.val=tmp.val
        if parent==root:
          parent.right=tmp.right
        else:
          parent.left=tmp.right
    return root

File: 卡码/代码随想录/1/06-二叉树/Go/06-二叉搜索树修改与构造/LeetCode_0450.go
//450. 删除二叉搜索树中的节点
package main

import (
	. "binaryTree/types"
)

func deleteNode(root *TreeNode, key int) *TreeNode {
	if root==nil {
		return nil
	}
	if key<root.Val {
		root.Left=deleteNode(root.Left,key)
	}else if key>root.Val {
		root.Right=deleteNode(root.Right,key)
	}else{
		if root.Left==nil&&root.Right==nil {
			return nil
		}else if root.Left!=nil&&root.Right==nil {
			return root.Left
		}else if root.Left==nil&&root.Right!=nil {
			return root.Right
		}else{
			parent:=root
			tmp:=root.Right
			for tmp.Left!=nil {
				parent=tmp
				tmp=tmp.Left
			}
			root.Val=tmp.Val
			if parent==root {
				parent.Right=tmp.Right
			}else{
				parent.Left=tmp.Right
			}
		}
	}
	return root
}
